sherpa/astro/instrument.py

```diff
@@ -1400,48 +1413,242 @@ def create_non_delta_rmf(rmflo, rmfhi, fname, offset=1,
+@dataclass
+class RMFMatrix:
+    """Raw RMF data"""
+
+    matrix: numpy.ndarray
+    """The matrix as a 2D array (X axis is channels, Y axis is energy)"""
+    channels: EvaluationSpace1D
+    """The channel values. This must be a non-integrated axis."""
+    energies: EvaluationSpace1D
+    """The energy values. This must be an integrated axis."""
+
+    def __post_init__(self) -> None:
+        if self.matrix.ndim != 2:
+            raise ValueError("matrix must be 2D")
+
+        if self.channels.is_integrated:
+            raise ValueError("channels axis must not be integrated")
+
+        if not self.energies.is_integrated:
+            raise ValueError("energies axis must be integrated")
+
+        nenergy, nchan = self.matrix.shape
+        if self.channels.x_axis.size != nchan:
+            raise ValueError("channels and matrix mis-match")
+
+        if self.energies.x_axis.size != nenergy:
+            raise ValueError("channels and matrix mis-match")
+
+
+def rmf_to_matrix(rmf: Union[DataRMF, RMF1D]) -> RMFMatrix:
+    """Convert a RMF to a matrix (2D image).
+
+    .. versionadded:: 4.16.0
+
+    Parameters
+    ----------
+    rmf : DataRMF or RMF1D
+       The RMF instance.
+
+    Returns
+    -------
+    info : RMFMatrix
+       The matrix as a 2D array (X axis is channels and Y axis is
+       energy, and the channel and energy axes.
+
+    """
+
+    if not isinstance(rmf, (DataRMF, RMF1D)):
+        raise ValueError("not a rmf")
+
+    # Create an image of size
+    #    nx = number of channels
+    #    ny = number of energy bine
+    #
+    nchans = rmf.detchans
+    nenergy = rmf.energ_lo.size
+    matrix = numpy.zeros((nenergy, nchans), dtype=rmf.matrix.dtype)
+
+    # Loop through each energy bin and add in the data, which is split
+    # into n_grp chunks, each starting at f_chan (with 1 being the
+    # first element of this row). The RMF has removed excess data -
+    # that is, rows with 0 groups and flattening out a 2D array for
+    # the n_chan/f_chan values - which makes the reconstruction a
+    # little messy.
+    #
+    matrix_start = 0
+    chan_idx = 0
+    for energy_idx, n_grp in enumerate(rmf.n_grp):
+        # Loop through the groups for this energy
+        for _ in range(n_grp):
+            # Need to convert from 1-based (f_chan) numbering
+            # (although this actually depends on the offset value) and
+            # to convert from numpy.uint64 (since <int> + <uint64>
+            # tends to get converted to a float).
+            #
+            start = rmf.f_chan[chan_idx].astype(int) - int(rmf.offset)
+            nchan = rmf.n_chan[chan_idx].astype(int)
+            end = start + nchan
+            matrix_end = matrix_start + nchan
+            matrix[energy_idx, start:end] = rmf.matrix[matrix_start:matrix_end]
+            matrix_start = matrix_end
+            chan_idx += 1
+
+    # TODO: Is the offset handling correct here?
+    channels = numpy.arange(rmf.offset, rmf.offset + nchans,
```

Comment:
I'm confused every time. Is offset the TLMIN header keyword?

Reply:
Yes. Conceptually I understand that we can have an offset (normally to swap between 1-based and 0-based but technically could be > 1) but whenever it comes to dealing with the code I always end up messing it up and I rely on the existing tests to catch my stupidity....

Which leads to the comment on line 1584